File: ios-client/Services/API/APIClient.swift
import Foundation

enum APIClientError: LocalizedError {
    case invalidBaseURL
    case invalidResponse
    case server(statusCode: Int, message: String)
    case decoding(underlying: Error)

    var errorDescription: String? {
        switch self {
        case .invalidBaseURL:
            return "Invalid backend URL."
        case .invalidResponse:
            return "Unexpected server response."
        case .server(_, let message):
            return message
        case .decoding:
            return "The server response could not be decoded."
        }
    }

    var statusCode: Int? {
        guard case .server(let statusCode, _) = self else {
            return nil
        }
        return statusCode
    }
}

final class APIClient {
    private let session: URLSession
    private let decoder: JSONDecoder
    private let encoder: JSONEncoder

    init(session: URLSession = .shared) {
        self.session = session

        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        self.decoder = decoder

        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        self.encoder = encoder
    }

    func login(
        baseURL: URL,
        username: String,
        password: String,
        installationID: String? = nil,
        proofTicket: String? = nil
    ) async throws -> LoginResponse {
        try await send(
            endpoint: .init(
                path: "/auth/login",
                method: "POST",
                body: try encoder.encode(
                    LoginRequest(
                        username: username,
                        password: password,
                        installationID: installationID,
                        proofTicket: proofTicket
                    )
                )
            ),
            baseURL: baseURL,
            token: nil
        )
    }

    func requestAppProofChallenge(
        baseURL: URL,
        platform: String,
        purpose: String,
        installationID: String,
        username: String? = nil,
        appVersion: String? = nil,
        osVersion: String? = nil
    ) async throws -> AppProofChallengeResponse {
        try await send(
            endpoint: .init(
                path: "/auth/app-proof/challenge",
                method: "POST",
                body: try encoder.encode(
                    AppProofChallengeRequest(
                        platform: platform,
                        purpose: purpose,
                        installationID: installationID,
                        username: username,
                        appVersion: appVersion,
                        osVersion: osVersion
                    )
                )
            ),
            baseURL: baseURL,
            token: nil
        )
    }

    func verifyIOSAppProof(
        baseURL: URL,
        mode: String,
        challengeID: String,
        installationID: String,
        keyID: String,
        attestationObject: String? = nil,
        assertionObject: String? = nil,
        receipt: String? = nil,
        appVersion: String? = nil,
        osVersion: String? = nil
    ) async throws -> AppProofVerifyResponse {
        try await send(
            endpoint: .init(
                path: "/auth/app-proof/verify/ios",
                method: "POST",
                body: try encoder.encode(
                    IOSAppProofVerifyRequest(
                        mode: mode,
                        challengeID: challengeID,
                        installationID: installationID,
                        keyID: keyID,
                        attestationObject: attestationObject,
                        assertionObject: assertionObject,
                        receipt: receipt,
                        appVersion: appVersion,
                        osVersion: osVersion
                    )
                )
            ),
            baseURL: baseURL,
            token: nil
        )
    }

    func fetchMyDevices(baseURL: URL, token: String, installationID: String? = nil) async throws -> [MyDevice] {
        try await send(
            endpoint: .init(path: "/me/devices", method: "GET"),
            baseURL: baseURL,
            token: token,
            installationID: installationID
        )
    }

    func wakeDevice(hostID: String, baseURL: URL, token: String, installationID: String? = nil) async throws -> MeWakeResponse {
        try await send(
            endpoint: .init(path: "/me/devices/\(hostID)/wake", method: "POST"),
            baseURL: baseURL,
            token: token,
            installationID: installationID
        )
    }

    func updateDevicePreferences(
        hostID: String,
        isFavorite: Bool? = nil,
        sortOrder: Int? = nil,
        baseURL: URL,
        token: String,
        installationID: String? = nil
    ) async throws -> MyDevice {
        try await send(
            endpoint: .init(
                path: "/me/devices/\(hostID)/preferences",
                method: "PATCH",
                body: try encoder.encode(
                    DevicePreferencesUpdateRequest(
                        isFavorite: isFavorite,
                        sortOrder: sortOrder
                    )
                )
            ),
            baseURL: baseURL,
            token: token,
            installationID: installationID
        )
    }

    func requestShutdownPoke(
        hostID: String,
        message: String?,
        baseURL: URL,
        token: String,
        installationID: String? = nil
    ) async throws -> ShutdownPoke {
        try await send(
            endpoint: .init(
                path: "/me/devices/\(hostID)/shutdown-poke",
                method: "POST",
                body: try encoder.encode(ShutdownPokeCreateRequest(message: message?.isEmpty == true ? nil : message))
            ),
            baseURL: baseURL,
            token: token,
            installationID: installationID
        )
    }

    func registerAPNSDevice(
        installationID: String,
        token deviceToken: String,
        appBundleID: String,
        environment: APNSEnvironment,
        baseURL: URL,
        authToken: String,
        authInstallationID: String? = nil
    ) async throws -> NotificationDeviceRegistration {
        try await send(
            endpoint: .init(
                path: "/me/notification-devices/apns",
                method: "POST",
                body: try encoder.encode(
                    APNSDeviceRegistrationRequest(
                        installationID: installationID,
                        token: deviceToken,
                        appBundleID: appBundleID,
                        environment: environment
                    )
                )
            ),
            baseURL: baseURL,
            token: authToken,
            installationID: authInstallationID
        )
    }

    func deleteAPNSDevice(
        installationID: String,
        baseURL: URL,
        authToken: String,
        authInstallationID: String? = nil
    ) async throws {
        try await sendWithoutResponse(
            endpoint: .init(
                path: "/me/notification-devices/apns/\(installationID)",
                method: "DELETE"
            ),
            baseURL: baseURL,
            token: authToken,
            installationID: authInstallationID
        )
    }

    func fetchAdminActivityEvents(
        baseURL: URL,
        token: String,
        cursor: Int?,
        limit: Int,
        typeFilter: String?,
        installationID: String? = nil
    ) async throws -> [ActivityEvent] {
        var queryItems = [URLQueryItem(name: "limit", value: String(limit))]
        if let cursor {
            queryItems.append(URLQueryItem(name: "cursor", value: String(cursor)))
        }
        if let typeFilter, !typeFilter.isEmpty {
            queryItems.append(URLQueryItem(name: "type", value: typeFilter))
        }

        return try await send(
            endpoint: .init(
                path: "/admin/mobile/events",
                method: "GET",
                queryItems: queryItems
            ),
            baseURL: baseURL,
            token: token,
            installationID: installationID
        )
    }

    func markShutdownPokeSeen(
        pokeID: String,
        baseURL: URL,
        token: String,
        installationID: String? = nil
    ) async throws -> ShutdownPoke {
        try await send(
            endpoint: .init(
                path: "/admin/shutdown-pokes/\(pokeID)/seen",
                method: "POST"
            ),
            baseURL: baseURL,
            token: token,
            installationID: installationID
        )
    }

    func markShutdownPokeResolved(
        pokeID: String,
        baseURL: URL,
        token: String,
        installationID: String? = nil
    ) async throws -> ShutdownPoke {
        try await send(
            endpoint: .init(
                path: "/admin/shutdown-pokes/\(pokeID)/resolve",
                method: "POST"
            ),
            baseURL: baseURL,
            token: token,
            installationID: installationID
        )
    }

    func normalizedBaseURL(from rawValue: String) throws -> URL {
        let trimmed = rawValue.trimmingCharacters(in: .whitespacesAndNewlines)
        guard var components = URLComponents(string: trimmed),
              let scheme = components.scheme?.lowercased(),
              ["http", "https"].contains(scheme),
              components.host != nil else {
            throw APIClientError.invalidBaseURL
        }
        if components.path.hasSuffix("/") {
            components.path.removeLast()
        }
        guard let url = components.url else {
            throw APIClientError.invalidBaseURL
        }
        return url
    }

    private func send<Response: Decodable>(
        endpoint: Endpoint,
        baseURL: URL,
        token: String?,
        installationID: String? = nil
    ) async throws -> Response {
        let request = try buildRequest(endpoint: endpoint, baseURL: baseURL, token: token, installationID: installationID)
        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw APIClientError.invalidResponse
        }

        guard (200..<300).contains(httpResponse.statusCode) else {
            throw APIClientError.server(
                statusCode: httpResponse.statusCode,
                message: decodeErrorMessage(from: data, statusCode: httpResponse.statusCode)
            )
        }

        do {
            return try decoder.decode(Response.self, from: data)
        } catch {
            throw APIClientError.decoding(underlying: error)
        }
    }

    private func sendWithoutResponse(
        endpoint: Endpoint,
        baseURL: URL,
        token: String?,
        installationID: String? = nil
    ) async throws {
        let request = try buildRequest(endpoint: endpoint, baseURL: baseURL, token: token, installationID: installationID)
        let (_, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw APIClientError.invalidResponse
        }

        guard (200..<300).contains(httpResponse.statusCode) else {
            throw APIClientError.server(
                statusCode: httpResponse.statusCode,
                message: HTTPURLResponse.localizedString(forStatusCode: httpResponse.statusCode)
            )
        }
    }

    private func buildRequest(
        endpoint: Endpoint,
        baseURL: URL,
        token: String?,
        installationID: String? = nil
    ) throws -> URLRequest {
        guard var components = URLComponents(url: baseURL, resolvingAgainstBaseURL: false) else {
            throw APIClientError.invalidBaseURL
        }

        let basePath = components.path.hasSuffix("/") ? String(components.path.dropLast()) : components.path
        components.path = basePath + endpoint.path
        if !endpoint.queryItems.isEmpty {
            components.queryItems = endpoint.queryItems
        }

        guard let url = components.url else {
            throw APIClientError.invalidBaseURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = endpoint.method
        request.httpBody = endpoint.body

        if endpoint.body != nil {
            request.setValue("application/json; charset=utf-8", forHTTPHeaderField: "Content-Type")
        }
        if let token {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        if let installationID, !installationID.isEmpty {
            request.setValue(installationID, forHTTPHeaderField: "X-WFF-Installation-ID")
        }

        return request
    }

    private func decodeErrorMessage(from data: Data, statusCode: Int) -> String {
        if let payload = try? decoder.decode(APIErrorEnvelope.self, from: data),
           let detail = payload.detail,
           !detail.isEmpty {
            return detail
        }
        if let raw = String(data: data, encoding: .utf8)?
            .trimmingCharacters(in: .whitespacesAndNewlines),
           !raw.isEmpty {
            return raw
        }
        return HTTPURLResponse.localizedString(forStatusCode: statusCode)
    }
}

private struct Endpoint {
    let path: String
    let method: String
    var queryItems: [URLQueryItem] = []
    var body: Data? = nil
}

private struct APIErrorEnvelope: Decodable {
    let detail: String?
}

File: android-client/app/src/main/java/com/wakefromfar/wolrelay/data/ApiClient.kt
package com.wakefromfar.wolrelay.data

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

class ApiClient(
    private val client: OkHttpClient = OkHttpClient.Builder().build(),
    private val json: Json = Json {
        ignoreUnknownKeys = true
        explicitNulls = false
    },
) {
    suspend fun login(
        baseUrl: String,
        username: String,
        password: String,
        installationId: String? = null,
        proofTicket: String? = null,
    ): LoginResponse = withContext(Dispatchers.IO) {
        val requestPayload = json.encodeToString(
            LoginRequest(
                username = username,
                password = password,
                installation_id = installationId,
                proof_ticket = proofTicket,
            ),
        )
        val request = Request.Builder()
            .url("${normalizeBaseUrl(baseUrl)}/auth/login")
            .post(requestPayload.toRequestBody(JSON_MEDIA_TYPE))
            .build()

        client.newCall(request).execute().use { response ->
            val responseBody = response.body?.string().orEmpty()
            if (!response.isSuccessful) {
                throw ApiException("Login failed (${response.code})", statusCode = response.code)
            }
            json.decodeFromString<LoginResponse>(responseBody)
        }
    }

    suspend fun requestAppProofChallenge(
        baseUrl: String,
        platform: String,
        purpose: String,
        installationId: String,
        username: String? = null,
        appVersion: String? = null,
        osVersion: String? = null,
    ): AppProofChallengeResponse = withContext(Dispatchers.IO) {
        val payload = json.encodeToString(
            AppProofChallengeRequest(
                platform = platform,
                purpose = purpose,
                installation_id = installationId,
                username = username,
                app_version = appVersion,
                os_version = osVersion,
            ),
        )
        val request = Request.Builder()
            .url("${normalizeBaseUrl(baseUrl)}/auth/app-proof/challenge")
            .post(payload.toRequestBody(JSON_MEDIA_TYPE))
            .build()

        client.newCall(request).execute().use { response ->
            val responseBody = response.body?.string().orEmpty()
            if (!response.isSuccessful) {
                throw ApiException(
                    errorFromResponse("App proof challenge failed", response.code, responseBody),
                    statusCode = response.code,
                )
            }
            json.decodeFromString<AppProofChallengeResponse>(responseBody)
        }
    }

    suspend fun verifyAndroidAppProof(
        baseUrl: String,
        challengeId: String,
        installationId: String,
        requestHash: String,
        integrityToken: String,
        appVersion: String? = null,
        osVersion: String? = null,
    ): AppProofVerifyResponse = withContext(Dispatchers.IO) {
        val payload = json.encodeToString(
            AndroidAppProofVerifyRequest(
                challenge_id = challengeId,
                installation_id = installationId,
                request_hash = requestHash,
                integrity_token = integrityToken,
                app_version = appVersion,
                os_version = osVersion,
            ),
        )
        val request = Request.Builder()
            .url("${normalizeBaseUrl(baseUrl)}/auth/app-proof/verify/android")
            .post(payload.toRequestBody(JSON_MEDIA_TYPE))
            .build()

        client.newCall(request).execute().use { response ->
            val responseBody = response.body?.string().orEmpty()
            if (!response.isSuccessful) {
                throw ApiException(
                    errorFromResponse("App proof verification failed", response.code, responseBody),
                    statusCode = response.code,
                )
            }
            json.decodeFromString<AppProofVerifyResponse>(responseBody)
        }
    }

    suspend fun claimOnboarding(baseUrl: String, token: String, password: String): OnboardingClaimResponse =
        withContext(Dispatchers.IO) {
            val payload = json.encodeToString(OnboardingClaimRequest(token = token, password = password))
            val request = Request.Builder()
                .url("${normalizeBaseUrl(baseUrl)}/onboarding/claim")
                .post(payload.toRequestBody(JSON_MEDIA_TYPE))
                .build()

            client.newCall(request).execute().use { response ->
                val responseBody = response.body?.string().orEmpty()
                if (!response.isSuccessful) {
                    throw ApiException(
                        errorFromResponse("Claim failed", response.code, responseBody),
                        statusCode = response.code,
                    )
                }
                json.decodeFromString<OnboardingClaimResponse>(responseBody)
            }
        }

    suspend fun listMyDevices(baseUrl: String, token: String, installationId: String? = null): List<MyDeviceDto> =
        withContext(Dispatchers.IO) {
        val request = authenticatedRequestBuilder(
            token = token,
            installationId = installationId,
        )
            .url("${normalizeBaseUrl(baseUrl)}/me/devices")
            .get()
            .build()

        client.newCall(request).execute().use { response ->
            val responseBody = response.body?.string().orEmpty()
            if (!response.isSuccessful) {
                throw ApiException(
                    errorFromResponse("Loading devices failed", response.code, responseBody),
                    statusCode = response.code,
                )
            }
            json.decodeFromString<List<MyDeviceDto>>(responseBody)
        }
    }

    suspend fun wakeDevice(
        baseUrl: String,
        token: String,
        hostId: String,
        installationId: String? = null,
    ): MeWakeResponse = withContext(Dispatchers.IO) {
        val request = authenticatedRequestBuilder(
            token = token,
            installationId = installationId,
        )
            .url("${normalizeBaseUrl(baseUrl)}/me/devices/$hostId/wake")
            .post("".toRequestBody(null))
            .build()

        client.newCall(request).execute().use { response ->
            val responseBody = response.body?.string().orEmpty()
            if (!response.isSuccessful) {
                throw ApiException(
                    errorFromResponse("Wake failed", response.code, responseBody),
                    statusCode = response.code,
                )
            }
            json.decodeFromString<MeWakeResponse>(responseBody)
        }
    }

    suspend fun updateMyDevicePreferences(
        baseUrl: String,
        token: String,
        hostId: String,
        isFavorite: Boolean? = null,
        sortOrder: Int? = null,
        installationId: String? = null,
    ): MyDeviceDto = withContext(Dispatchers.IO) {
        val payload = json.encodeToString(
            MyDevicePreferencesUpdateRequest(
                is_favorite = isFavorite,
                sort_order = sortOrder,
            ),
        )
        val request = authenticatedRequestBuilder(
            token = token,
            installationId = installationId,
        )
            .url("${normalizeBaseUrl(baseUrl)}/me/devices/$hostId/preferences")
            .patch(payload.toRequestBody(JSON_MEDIA_TYPE))
            .build()

        client.newCall(request).execute().use { response ->
            val responseBody = response.body?.string().orEmpty()
            if (!response.isSuccessful) {
                throw ApiException(
                    errorFromResponse("Updating device preferences failed", response.code, responseBody),
                    statusCode = response.code,
                )
            }
            json.decodeFromString<MyDeviceDto>(responseBody)
        }
    }

    suspend fun listAdminEvents(
        baseUrl: String,
        token: String,
        cursor: Int? = null,
        limit: Int = 50,
        typeFilter: String? = "wake,poke",
        installationId: String? = null,
    ): List<ActivityEventDto> = withContext(Dispatchers.IO) {
        val urlBuilder = "${normalizeBaseUrl(baseUrl)}/admin/mobile/events".toHttpUrl().newBuilder()
            .addQueryParameter("limit", limit.toString())
        if (cursor != null) {
            urlBuilder.addQueryParameter("cursor", cursor.toString())
        }
        if (!typeFilter.isNullOrBlank()) {
            urlBuilder.addQueryParameter("type", typeFilter)
        }
        val request = authenticatedRequestBuilder(
            token = token,
            installationId = installationId,
        )
            .url(urlBuilder.build())
            .get()
            .build()

        client.newCall(request).execute().use { response ->
            val responseBody = response.body?.string().orEmpty()
            if (!response.isSuccessful) {
                throw ApiException(
                    errorFromResponse("Loading admin events failed", response.code, responseBody),
                    statusCode = response.code,
                )
            }
            json.decodeFromString<List<ActivityEventDto>>(responseBody)
        }
    }

    suspend fun requestShutdownPoke(
        baseUrl: String,
        token: String,
        hostId: String,
        message: String? = null,
        installationId: String? = null,
    ): ShutdownPokeDto = withContext(Dispatchers.IO) {
        val payload = json.encodeToString(
            ShutdownPokeCreateRequest(
                message = message?.trim()?.ifBlank { null },
            ),
        )
        val request = authenticatedRequestBuilder(
            token = token,
            installationId = installationId,
        )
            .url("${normalizeBaseUrl(baseUrl)}/me/devices/$hostId/shutdown-poke")
            .post(payload.toRequestBody(JSON_MEDIA_TYPE))
            .build()

        client.newCall(request).execute().use { response ->
            val responseBody = response.body?.string().orEmpty()
            if (!response.isSuccessful) {
                throw ApiException(
                    errorFromResponse("Shutdown request failed", response.code, responseBody),
                    statusCode = response.code,
                )
            }
            json.decodeFromString<ShutdownPokeDto>(responseBody)
        }
    }

    suspend fun markShutdownPokeSeen(
        baseUrl: String,
        token: String,
        pokeId: String,
        installationId: String? = null,
    ): ShutdownPokeDto = withContext(Dispatchers.IO) {
        val request = authenticatedRequestBuilder(
            token = token,
            installationId = installationId,
        )
            .url("${normalizeBaseUrl(baseUrl)}/admin/shutdown-pokes/$pokeId/seen")
            .post("".toRequestBody(null))
            .build()

        client.newCall(request).execute().use { response ->
            val responseBody = response.body?.string().orEmpty()
            if (!response.isSuccessful) {
                throw ApiException(
                    errorFromResponse("Marking shutdown request seen failed", response.code, responseBody),
                    statusCode = response.code,
                )
            }
            json.decodeFromString<ShutdownPokeDto>(responseBody)
        }
    }

    suspend fun markShutdownPokeResolved(
        baseUrl: String,
        token: String,
        pokeId: String,
        installationId: String? = null,
    ): ShutdownPokeDto = withContext(Dispatchers.IO) {
        val request = authenticatedRequestBuilder(
            token = token,
            installationId = installationId,
        )
            .url("${normalizeBaseUrl(baseUrl)}/admin/shutdown-pokes/$pokeId/resolve")
            .post("".toRequestBody(null))
            .build()

        client.newCall(request).execute().use { response ->
            val responseBody = response.body?.string().orEmpty()
            if (!response.isSuccessful) {
                throw ApiException(
                    errorFromResponse("Resolving shutdown request failed", response.code, responseBody),
                    statusCode = response.code,
                )
            }
            json.decodeFromString<ShutdownPokeDto>(responseBody)
        }
    }

    private fun normalizeBaseUrl(baseUrl: String): String = baseUrl.trim().trimEnd('/')

    private fun authenticatedRequestBuilder(
        token: String,
        installationId: String? = null,
    ): Request.Builder {
        val builder = Request.Builder()
            .addHeader("Authorization", "Bearer $token")
        if (!installationId.isNullOrBlank()) {
            builder.addHeader("X-WFF-Installation-ID", installationId)
        }
        return builder
    }

    private fun errorFromResponse(prefix: String, code: Int, responseBody: String): String {
        val compact = responseBody.replace("\n", " ").trim().take(200)
        return if (compact.isBlank()) "$prefix ($code)" else "$prefix ($code): $compact"
    }

    private companion object {
        val JSON_MEDIA_TYPE = "application/json; charset=utf-8".toMediaType()
    }
}

class ApiException(
    message: String,
    val statusCode: Int? = null,
) : RuntimeException(message)
